### PR TITLE
Done

### PR DESCRIPTION
**Pull request by vovkandrew**

### CODE REVIEW

File: src/main/java/cinema/controllers/AuthenticationController.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import cinema.model.dto.UserRequestDto;
+import cinema.service.AuthenticationService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/register")
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/controllers/AuthenticationController.java]: let's move it to method, probably you'll have @RequestMapping("/login") method in future in this particular controller

File: src/main/java/cinema/controllers/MovieSessionController.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/moviesessions")
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/controllers/MovieSessionController.java]:
```suggestion
@RequestMapping("/movie-sessions")
```

File: src/main/java/cinema/controllers/OrderController.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    @PostMapping("/complete")
+    public void completeOrder(@RequestBody UserRequestDto userRequestDto) {
+        orderService.completeOrder(
+                shoppingCartService.getByUser(
+                        userService.findByEmail(userRequestDto.getEmail())).getTickets(),
+                userService.findByEmail(userRequestDto.getEmail()));
+    }
+
+    @GetMapping("/orders")
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/controllers/OrderController.java]:
```suggestion
    @GetMapping
```
image

File: src/main/java/cinema/controllers/ShoppingCartController.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import cinema.service.UserService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/shoppingcarts")
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/controllers/ShoppingCartController.java]:
```suggestion
@RequestMapping("/shopping-carts")
```

File: src/main/java/cinema/dao/impl/CinemaHallDaoImpl.java [hunk 5]
```diff
@@ -57,4 +58,18 @@ public CinemaHall add(CinemaHall cinemaHall) {
...
     }
+
+    @Override
+    public CinemaHall getById(Long id) {
+        try (Session session = sessionFactory.openSession()) {
+            Query<CinemaHall> query = session.createQuery(
+                    "from CinemaHall where id = :id", CinemaHall.class);
+            query.setParameter("id", id);
+            LOGGER.info("Cinema hall with id " + id + " has been found in the database");
+            return query.uniqueResult();
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/dao/impl/CinemaHallDaoImpl.java]: session.get
```suggestion
           return session.get(CinemaHall.class, id);
```
try this one here and everywhere(each dao) where you are getting entity by id

File: src/main/java/cinema/model/mapper/CinemaHallMapper.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import cinema.model.dto.CinemaHallRequestDto;
+import cinema.model.dto.CinemaHallResponseDto;
+import org.springframework.stereotype.Component;
+
+@Component
+public class CinemaHallMapper {
+    public CinemaHall getCinemaHallFromCinemaHallRequestDto(
+            CinemaHallRequestDto cinemaHallRequestDto) {
+        return new CinemaHall(cinemaHallRequestDto.getCapacity(),
+                cinemaHallRequestDto.getDescription());
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/model/mapper/CinemaHallMapper.java]: prefer setters over long constructors. Even if it takes more lines of code it's more readable and easier to refactor

Comment (dmytro-hryhoruk) [reply src/main/java/cinema/model/mapper/CinemaHallMapper.java]: let's better remove such constructors and remove them from theirs classes

File: src/main/java/cinema/model/mapper/MovieSessionMapper.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    @Autowired
+    private CinemaHallService cinemaHallService;
+
+    public MovieSession getMovieSessionFromMovieSessionRequestDto(
+            MovieSessionRequestDto movieSessionRequestDto) {
+        Movie movie = movieService.getById(movieSessionRequestDto.getMovieId());
+        CinemaHall cinemaHall = cinemaHallService.getById(
+                movieSessionRequestDto.getCinemaHallId());
+        LocalDateTime showtime = LocalDateTime.parse(movieSessionRequestDto.getTime());
+        return new MovieSession(movie,cinemaHall,showtime);
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/model/mapper/MovieSessionMapper.java]:
```suggestion
        return new MovieSession(movie, cinemaHall, showtime);
```
1) don't forget to run mvn clean package before pushing something to git;
2)let's use setters))

Comment (vovkandrew) [reply src/main/java/cinema/model/mapper/MovieSessionMapper.java]: I always do that on a regular basis, but you always find one single syntax mistake :)

File: src/main/java/cinema/controllers/CinemaHallController.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.stream.Collectors;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/cinemahalls")
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/controllers/CinemaHallController.java]:
```suggestion
@RequestMapping("/cinema-halls")
```

File: src/main/java/cinema/controllers/OrderController.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+    @PostMapping("/complete")
+    public void completeOrder(@RequestBody UserRequestDto userRequestDto) {
+        orderService.completeOrder(
+                shoppingCartService.getByUser(
+                        userService.findByEmail(userRequestDto.getEmail())).getTickets(),
+                userService.findByEmail(userRequestDto.getEmail()));
+    }
+
+    @GetMapping("/all")
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/controllers/OrderController.java]:
```suggestion
    @GetMapping
```

File: src/main/java/cinema/controllers/ShoppingCartController.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    private ShoppingCartMapper shoppingCartMapper;
+
+    @PostMapping("/addmoviesession")
+    public void addMovieSession(@RequestParam Long sessionId, @RequestParam Long userId) {
+        shoppingCartService.addSession(
+                movieSessionService.getById(sessionId),
+                userService.getById(userId));
+    }
+
+    @GetMapping("/byuser")
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/controllers/ShoppingCartController.java]:
```suggestion
    @GetMapping("/by-user")
```

File: src/main/java/cinema/controllers/ShoppingCartController.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    @Autowired
+    private MovieSessionService movieSessionService;
+
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    private ShoppingCartMapper shoppingCartMapper;
+
+    @PostMapping("/addmoviesession")
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/controllers/ShoppingCartController.java]:
```suggestion
    @PostMapping("/add-movie-session")
```

File: src/main/java/cinema/controllers/UserController.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+@RestController
+@RequestMapping("/user")
+public class UserController {
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    private UserMapper userMapper;
+
+    @GetMapping("/byemail")
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/controllers/UserController.java]:
```suggestion
    @GetMapping("/by-email")
```

File: src/main/java/cinema/model/mapper/TicketMapper.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+@Component
+public class TicketMapper {
+    @Autowired
+    private TicketService ticketService;
+
+    public TicketResponseDto getTicketResponseDtoFromTicket(Ticket ticket) {
+        return new TicketResponseDto(ticket.getSession().getMovie().getTitle(),
+                ticket.getSession().getCinemaHall().getId(),
+                ticket.getSession().getTime().toString());
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/model/mapper/TicketMapper.java]: let's use setters over long constructors( check this in whole project)

**Review by dmytro-hryhoruk — COMMENTED** [review]
Внесит правки і мерджи)

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**